### PR TITLE
new chart: netstat

### PR DESCRIPTION
**Pull request by Wing924**
It collect tcp connection state from netstat command.

### CODE REVIEW

File: python.d/netstat.chart.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        raw_data = self._get_raw_data()
+        if not raw_data:
+            return None
+        ptn = re.compile(r'\s+')
+        data = {}
+        for key in ['ESTABLISHED', 'SYN_SENT', 'SYN_RECV', 'FIN_WAIT1', 'FIN_WAIT2', 'TIME_WAIT', 'CLOSE', 'CLOSE_WAIT', 'LISTEN', 'CLOSING']:
+            data[key] = 0
+
+        for line in raw_data:
+            tk = ptn.split(line)
```
Comment (l2isbad) [inline python.d/netstat.chart.py]: Do we really need re package here? Looks like simple line.split() will do the work. Can you check it?

**Comment by l2isbad** [comment]
@Wing924 check codacy error

File: python.d/netstat.chart.py (hunk 1 above)
Comment (Wing924) [inline python.d/netstat.chart.py]: @l2isbad 
This is sample output form netstat:
```
tcp4       0      0  10.49.49.61.55709      40.100.155.2.443       ESTABLISHED
tcp4       0      0  10.49.49.61.55708      40.100.155.2.443       ESTABLISHED
tcp4      31      0  10.49.49.61.55679      192.30.255.113.443     CLOSE_WAIT
```
Results contain many spaces. Can you tell me how to split without regex?

Comment (l2isbad) [inline python.d/netstat.chart.py]:
```python
'hello world            and              goodbuy            !'.split()
['hello', 'world', 'and', 'goodbuy', '!']
```

File: python.d/netstat.chart.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+# Description: example netdata python.d module
+# Author: Pawel Krupa (paulfantom)
```
Comment (l2isbad) [inline python.d/netstat.chart.py]: really? 😄

**Comment by l2isbad** [comment]
thx @Wing924 
So it's only works for ipv4.

I am not an expert but looks like we can extract the same info from `/proc/net/tcp` which is much faster.

`man 5 proc`
```
       /proc/net/tcp
              Holds a dump of the TCP socket table.  Much of the information is not of use apart from debugging.   The  "sl"  value  is  the  kernel  hash  slot  for  the  socket,  the
              "local_address"  is  the local address and port number pair.  The "rem_address" is the remote address and port number pair (if connected).  "St" is the internal status of
              the socket.  The "tx_queue" and "rx_queue" are the outgoing and incoming data queue in terms of kernel memory usage.  The "tr",  "tm->when",  and  "rexmits"  fields  hold
              internal information of the kernel socket state and are useful only for debugging.  The "uid" field holds the effective UID of the creator of the socket
```

[St value](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/tcp_states.h?id=HEAD) (internal status of the socket)

@ktsaou ?

**Comment by ktsaou** [comment]
#2174

I will try to come up with a plugin to do this in C.
I haven't decided it yet how it should be done though:

1. A plugin inside netdata daemon
2. A feature of apps.plugin
3. something completely new

Ideally, I would like to have this:

1. count sockets globally (like this plugin)
2. count sockets per application (in apps.plugin, but the presentation is somewhat complex - we will need a chart per status with all the applications as dimensions)
3. count sockets of one or more interesting ports (i.e. generate a chart per port?)

For sure, we need an **application watchdog** in netdata. Something that will monitor everything about an application or a pid.

I think we should do this in `apps.plugin`... this will also allow us to have sockets per User and sockets per User Group.

What do you think?

**Comment by Wing924** [comment]
@ktsaou 
>I think we should do this in apps.plugin

I agree with it because it's the fastest way to do so and can provide more detailed info.

> count sockets globally (like this plugin)

I use this plugin and found many bugs in my system such as #2446. I think it is also important for other users.

> count sockets per application (in apps.plugin, but the presentation is somewhat complex - we will need a chart per status with all the applications as dimensions)

I think count established sockets and create 2 charts: connect from and connect to is useful and simple.
If destination address is self, it is  connect to, otherwise connect from.

> count sockets of one or more interesting ports (i.e. generate a chart per port?)

I think generating a chart per listening port is useful.

**Comment by Wing924** [comment]
@l2isbad I changed data source from netstat command to /proc/net/tcp. Can you review it?

**Comment by l2isbad** [comment]
@Wing924 i think it will be implemented by @ktsaou in C

**Comment by ktsaou** [comment]
Yes, I will. This is too heave to implemented in python.

**Pull request closed without merging**